unix, win: extend uv_udp_set_membership for IPv6 multicast.

snoj opened the pull request:
Added uv_udp_set_membership6.
Modified uv_udp_set_membership to call uv_udp_set_membership6
  when IPv6 is used.

Google groups discussion...not much there mainly easy access for myself. https://groups.google.com/forum/?fromgroups=#!topic/libuv/TxZmEnH2wP4

snoj commented:
Ahhh shoot. Previous comments you guys put in are gone.

Anyway, made the changes that were recommended and refined the multicast binding. Now it doesn't send out on all interfaces if one is given. In order to do this though, I had to get the interface index which I did by extending uv_interface_address_s.

I'll start working on the Linux versions if there is nothing more.

txdv commented:
If you come to the irc channel right now, I will aid you in this task.

bnoordhuis commented on src/win/udp.c:
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+  
+  if(interface_addr) {
```
Style: space between keyword and paren.

bnoordhuis commented on src/win/udp.c:
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
+      uv__set_sys_error(handle->loop, WSAGetLastError());
+      return -1;
```
Fold into previous line.

bnoordhuis commented on src/win/udp.c:
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+
+    while(interfaces_count--) {
+      if(interfaces[interfaces_count].address.address4.sin_family == AF_INET6) {
+        if(memcmp(&interfaces[interfaces_count].address.address6.sin6_addr, &interface_addr_n, sizeof interface_addr_n) == 0) {
```
Long line.

bnoordhuis commented on src/win/udp.c:
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  
+  if(interface_addr) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
```
Leaking interfaces?

bnoordhuis commented on src/win/util.c:
```diff
@@ -839,6 +839,9 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses_ptr,
       continue;
 
+    /* Add four bytes for the interface index */
+    uv_address_buf_size += 4;
```
Use sizeof(fieldname) here.

bnoordhuis replied:
By the way, is it necessary? It gets incremented with sizeof(uv_interface_address_t) further down below, right?

bnoordhuis commented on src/win/util.c:
```diff
@@ -901,6 +904,10 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses_ptr,
                                     FALSE);
+
+    uv_address->ifIndex = 0;
+    uv_address->ifIndex = (win_address->IfIndex) ? win_address->IfIndex : win_address->Ipv6IfIndex;
```
Superfluous assignment, superfluous parens.

bnoordhuis commented on include/uv.h:
```diff
@@ -1433,6 +1450,7 @@ struct uv_cpu_info_s {
 struct uv_interface_address_s {
   char* name;
   int is_internal;
+  ULONG ifIndex;
```
Use snake_case, not camelCase.

snoj commented:
Super special thanks to @txdv for fixing the bad code style!

bnoordhuis commented on src/unix/internal.h:
```diff
@@ -244,3 +244,6 @@ static void uv__update_time(uv_loop_t* loop) {
 #endif /* UV_UNIX_INTERNAL_H_ */
+
+/*  */
+#define UV_HANDLE_IPV6                          0x01000000
```
What's that doing there?

snoj replied:
A copy of UV_HANDLE_IPV6 from src/win/internal.h. It was otherwise not available on my linux build.

bnoordhuis replied:
Okay, but look closely _where_ you put it in src/unix/internal.h.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -524,6 +525,11 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
   struct ip_mreq mreq;
+
+  if (handle->flags & UV_HANDLE_IPV6) {
+    return uv_udp_set_membership6(handle, multicast_addr, interface_addr, membership);
```
If we're doing an implicit call then uv_udp_set_membership6 doesn't have to be public (and should be renamed to uv__udp_set_membership6 to indicate it's internal.)

bnoordhuis replied:
EDIT: it should probably be static as well.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
 }
 
+int uv_udp_set_membership6(uv_udp_t* handle, const char* multicast_addr,
+  const char* interface_addr, uv_membership membership) {
```
Arguments should line up, i.e.:

``` c
int uv_udp_set_membership6(uv_udp_t* handle,
                           const char* multicast_addr,
                           const char* interface_addr,
                           uv_membership membership
```

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  struct ipv6_mreq mreq;
+  struct in6_addr multicast_addr_n;
+  struct in6_addr interface_addr_n;
+
```
Style: No newline.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
+      uv__set_sys_error(handle->loop, UV_EINVAL);
+      return -1;
```
Style: fold into a single-line statement (`return uv__set_sys_error(handle->loop, UV_EINVAL);`) and drop the braces.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+      return -1;
+    }
+
+    while (interfaces_count--) {
```
Micro-optimization and micro-style: use a `for (i = 0; i < interfaces_count; i++)` loop here.

Micro-optimization: CPU prefetchers are good at recognizing memory access patterns in ascending order but generally not when they're in descending order.

Micro-style: Prefer not mutating variables. It's a kind of future-proofing; if someone changes the code later on and assumes interfaces_count is factual, things will go badly wrong.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+
+    while (interfaces_count--) {
+      if (interfaces[interfaces_count].address.address4.sin_family == AF_INET6) {
+        if (memcmp(&interfaces[interfaces_count].address.address6.sin6_addr, &interface_addr_n, sizeof interface_addr_n) == 0) {
```
Style: Long line. Wrap at 80 characters max. I believe I pointed that out last night too. :-/

bnoordhuis replied:
Also, using memcmp() here is potentially wrong. The struct may have holes in it that contain garbage.

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+
+  if (interface_addr) {
```
Style: interface_addr != NULL

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+    return -1;
+  }
+
+  if (setsockopt(handle->io_watcher.fd, IPPROTO_IPV6, optname, (void*) &mreq, sizeof mreq) == -1) {
```
Superfluous cast to void*.

txdv replied:
- space

bnoordhuis commented on test/test-udp-multicast-join6.c:
```diff
@@ -0,0 +1,140 @@
+
+static int close_cb_called;
+
+static uv_buf_t alloc_cb(uv_handle_t* handle, size_t suggested_size) {
```
Style: Two newlines between declarations and functions.

snoj committed:
unix, windows: extend uv_udp_set_membership for IPv6 multicast.

Add uv__udp_set_membership6.
Modify uv_udp_set_membership to call uv__udp_set_membership6
  when IPv6 is used.
Add if_index to struct uv_interface_address_s for IPv6 multicasting.
Add UV_HANDLE_IPV6 to unix/interna.h

Signed-off-by: snoj <josh@snoj.us>

snoj commented:
I believe the new forced commit takes care of most of the comments. I couldn't figure out a way compare ip addresses without more variables and conversions, perhaps I'll need to resort to that that.

diversario commented:
I've tried to use multicast with IPv6 address but I keep getting `addMembership EINVAL`. Does this PR address that?

snoj commented:
@diversario, When I started this, node/libuv did not support ipv6 multicast. I don't recall that EINVAL (see google group link at the top), but they may have changed the error codes since.

Since you're trying to use ipv6 multicasting, I think something like this would work for you. This code probably several version behind by now though. I or someone would need to review it.

saghul commented:
Sorry this was left hanging. Can yoou please rebase it against current master?

indutny commented on src/unix/linux/linux-core.c:
```diff
@@ -717,6 +717,7 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses,
 #else
   struct ifaddrs *addrs, *ent;
   char ip[INET6_ADDRSTRLEN];
+
```
Unnecessary whitespace change.

snoj commented:
@saghul I'll attempt to rebase and run tests by sometime this coming weekend. But with it being christmas and with all the family get togethers, I can't guarantee anything.

saghul commented:
No problem Josh!
On Dec 19, 2013 1:28 AM, "Josh Erickson" notifications@github.com wrote:

> @saghul https://github.com/saghul I'll attempt to rebase and run tests
> by sometime this coming weekend. But with it being christmas and with all
> the family get togethers, I can't guarantee anything.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/692#issuecomment-30895352
> .

snoj commented:
Definitely not happening this weekend. To many changes have happened since writing this.

saghul commented:
No need to hurry :-)
On Dec 21, 2013 7:12 AM, "Josh Erickson" notifications@github.com wrote:

> Definitely not happening this weekend. To many changes have happened since
> writing this.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/joyent/libuv/pull/692#issuecomment-31057751
> .

mprokopowicz commented:
Any progress on this?  Can't wait for got this in node release.
It's critical for my project :)

snoj commented:
Been working the last couple/few days with @mprokopowicz on this. Appears to be working, but when I rebased on the latest master, node broke. Tonight I'm hoping to rebase on the version node is using and going from there.

snoj commented:
Done? Code compiles and ipv6 multicast tests completes without errors.

!ping @saghul ?

indutny commented on src/unix/udp.c:
```diff
@@ -35,6 +35,10 @@
 static void uv__udp_recvmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static void uv__udp_sendmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static int uv__udp_maybe_deferred_bind(uv_udp_t* handle, int domain);
+static int uv__udp_set_membership6(uv_udp_t* handle, 
```
Whitespace at the end of the line

indutny commented on src/unix/udp.c:
```diff
@@ -35,6 +35,10 @@
 static int uv__udp_maybe_deferred_bind(uv_udp_t* handle, int domain);
+static int uv__udp_set_membership6(uv_udp_t* handle, 
+                                   const char* multicast_addr,
+                                   const char* interface_addr, 
```
Ditto

indutny commented on src/win/udp.c:
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  struct in6_addr multicast_addr_n;
+  struct in6_addr interface_addr_n;
+  uv_interface_address_t* interfaces;
+    
```
Ditto

indutny commented:
Got this on OSX:

```
../../src/unix/udp.c:549:15: error: use of undeclared identifier 'IPV6_ADD_MEMBERSHIP'
    optname = IPV6_ADD_MEMBERSHIP;
              ^
../../src/unix/udp.c:552:15: error: use of undeclared identifier 'IPV6_DROP_MEMBERSHIP'
    optname = IPV6_DROP_MEMBERSHIP;
              ^
```

indutny commented:
Perhaps it is called `IPV6_JOIN_GROUP` and `IPV6_LEAVE_GROUP` on osx?

indutny commented:
I'd suggest adding `UV__IPV6_JOIN_GROUP` and `UV__IPV6_LEAVE_GROUP` in `src/unix/internal.h` and setting them to the appropriate value

indutny commented on test/test-udp-multicast-join6.c:
```diff
@@ -0,0 +1,148 @@
+
+  ASSERT(0 == uv_ip6_addr("::1", TEST_PORT, &addr));
+
+  r = uv_udp_init(uv_default_loop(), &server);
```
Please do ASSERT(0 == ...) everywhere

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count) != 0)
+      return -1;
+      /*return uv__set_sys_error(handle->loop, UV_EINVAL);*/
```
This is a bad idea :)

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+  if (interface_addr != NULL) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count) != 0)
+      return -1;
```
Just return `-errno`

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+
+    if (mreq.ipv6mr_interface == 0) {
+      /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+      return -1;
```
Ditto

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+
+  if (uv_inet_pton(AF_INET6, multicast_addr, &multicast_addr_n) != 0) {
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Ditto

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+    }
+  }
+
+  if (uv_inet_pton(AF_INET6, multicast_addr, &multicast_addr_n) != 0) {
```
And no need in braces if there will be one line

indutny replied:
Also in all other places in diff.

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+    break;
+  default:
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Ditto

indutny replied:
Probably should be an assertion, or UV_ENOTSUP

indutny commented on src/unix/udp.c:
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
+
+  if (setsockopt(handle->io_watcher.fd, IPPROTO_IPV6, optname, &mreq, sizeof mreq) != 0) {
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Dito

indutny commented on src/win/udp.c:
```diff
@@ -539,7 +539,7 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
 
   /* If the socket is unbound, bind to inaddr_any. */
-  if (!(handle->flags & UV_HANDLE_BOUND)) {
+  if (!(handle->flags & UV_HANDLE_IPV6) && !(handle->flags & UV_HANDLE_BOUND)) {
```
`!(handle->flags & (UV_HANDLE_IPV6 | UV_HANDLE_BOUND))`

indutny commented on src/win/udp.c:
```diff
@@ -570,7 +570,7 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
       break;
     default:
-      return UV_EINVAL;
+      return UV__EINVAL;
```
Why?

indutny commented on src/win/udp.c:
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+  uv_interface_address_t* interfaces;
+    
+  if (!(handle->flags & UV_HANDLE_IPV6)) {
+	  return -1;
```
Ditto

indutny commented on src/win/udp.c:
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+    }
+
+    if (mreq.ipv6mr_interface == 0) {
+      return -1;
```
Should return some meaningful error here

indutny commented on src/win/udp.c:
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
+      optname = IPV6_DROP_MEMBERSHIP;
+      break;
+    default:
+      return -1;
```
And there

indutny commented:
As another note, I do get `EADDRNOTAVAIL` on OSX and test is failing.

indutny commented:
Please fix all that things and I will finish reviewing

indutny commented:
And generally it looks quite neat ;)

saghul commented:
The UV_HANDLE_IPV6 flag is only used for UDP, maybe you could rename it to something like UV_UDP_IPV6?

saghul commented:
I'd name the extra field you added to uv_interface_address_s just "index". Also, AFAIS it's only set on Linux and Windows. Don't other platforms need it as well?

saghul commented:
Instead of branching out to a ipv6 version of the function and continuing on the same otherwise, I'd split uv_udp_set_membership into 2: uv__udp_set_membership4/6.

PS: sorry for the brevity, I'm on mobile.

mprokopowicz commented:
Hi,

i was mailing with snoj and unfortunately he won't have time to finish this commit. Is any one here who can handle it?

saghul commented:
I'll take over.

mprokopowicz commented:
@saghul Awsome news :) Can U provide any estimates? (No pressure, just curious... )

saghul commented:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

On 03/14/2014 10:01 AM, Michał Prokopowicz wrote:

> @saghul https://github.com/saghul Awsome news :) Can U provide
> any estimates? (No pressure, just curious... )

We are _very_ close to 0.12. Ideally this would make it. IIRC the code
was almost  there, I'll take a stab at it this weekend if possible.

---

Saúl Ibarra Corretgé
bettercallsaghul.com

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1
Comment: Using GnuPG with Icedove - http://www.enigmail.net/

iQIcBAEBAgAGBQJTIsX3AAoJEEEOVVOum8BZLPcQAMQiRujwDiSNhPtRKDDw9rX+
e+tuOsJaMzW1rwf6COpFgxAKmuGmkm/U9LIRUBdD93v8RYYIPLk3bsp0XMTqCBa/
DGvxxZc9xYZJGp+QNXboPyjF1XWBOGflOWOmVHCmb0shABT+3KiKuB2jcv1r+9sV
iM4+/9HmliEcZROXp/mOykapTmLlVp7cH04CmBXuq2Y325TtuhBaAcxErfT1h89/
LpxPERuQFEqlsXYffPouQDf8CZEBb2ZfoK54K7Zh5vYnHa/38rjbr/WIPpEsRLvt
Yy3kvYD5IHeXqvNJ5MlsNQuh7Bd5mdt9TjB6nsJo+i3ieX393mwzPt8Hngqu07/n
L2pDHRKXI7Fpe9SNfb3PwfrKQjukymvtbhYdsTfNksGqM+HjEBFGbPv1HM3IC0RE
PGFqF0Tu0IxNBvxgnhFd9h5dRSQKD515SXmZXp5iCtcB3gFNxjRNjFIwaeyg3RcG
SOEU73qQhgC/zJ/ueWdu1C0FggDqUuu3+VWHcmQ8byv+/o0Ltpv7lG+CBgqgkXvh
HRy0SkAhyjtgvNfhDF+BvSbWnbmnKke24rZ+wNSKdAZea5c5uz/RlPc18+Q7EFbp
CLuG2b/BPnIhgLRlxOKv4IDnYBFCDABAutTLCJ1u8FY2EACGjNbMWQ1ACt0J6+di
tGuGgIt4wCLfMfVRsqSk
=YV1u
-----END PGP SIGNATURE-----

saghul commented:
For those interested, I reworked the Unix part here: https://github.com/saghul/libuv/tree/set-membership-v6

I'll send a proper PR for review once I'm done with the Windows part.

saghul commented:
Superseded by https://github.com/joyent/libuv/pull/1202. Lets continue the discussion there.
